src/lib/io/include/sol-memmap-storage.h

```diff
@@ -0,0 +1,273 @@
+ * behaviour is undefined)
+ * @param buffer buffer where result will be read into, according to its entry
+ * on map.
+ *
+ * return 0 on success, a negative number on failure
+ */
+int sol_memmap_read_raw(const char *name, struct sol_buffer *buffer);
+
+/**
+ * Add a map to internal list of available maps.
+ *
+ * @param map map to be add.
+ *
+ * @return 0 on success, a negative number on failure.
+ */
+int sol_memmap_add_map(const struct sol_memmap_map *map);
+
+/**
+ * Removes a previously added map from internal list of available maps.
+ *
+ * @param map map to be removed.
+ *
+ * @return 0 on success, a negative number on failure.
+ */
+int sol_memmap_remove_map(const struct sol_memmap_map *map);
+
+#define CREATE_BUFFER(_val, _empty) \
+    struct sol_buffer buf = SOL_BUFFER_INIT_FLAGS(_val, \
+    sizeof(*(_val)), SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED); \
+    buf.used = (_empty) ? 0 : sizeof(*(_val));
+
+static inline int
+sol_memmap_read_uint8(const char *name, uint8_t *value)
+{
+    CREATE_BUFFER(value, true);
+
+    return sol_memmap_read_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_write_uint8(const char *name, uint8_t value)
+{
+    CREATE_BUFFER(&value, false);
+
+    return sol_memmap_write_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_read_bool(const char *name, bool *value)
+{
+    CREATE_BUFFER(value, true);
+
+    return sol_memmap_read_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_write_bool(const char *name, bool value)
+{
+    CREATE_BUFFER(&value, false);
+
+    return sol_memmap_write_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_read_int32(const char *name, int32_t *value)
+{
+    CREATE_BUFFER(value, true);
+
+    return sol_memmap_read_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_write_int32(const char *name, int32_t value)
+{
+    CREATE_BUFFER(&value, false);
+
+    return sol_memmap_write_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_read_irange(const char *name, struct sol_irange *value)
+{
+    CREATE_BUFFER(value, true);
+
+    return sol_memmap_read_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_write_irange(const char *name, struct sol_irange *value)
+{
+    CREATE_BUFFER(value, false);
+
+    return sol_memmap_write_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_read_drange(const char *name, struct sol_drange *value)
+{
+    CREATE_BUFFER(value, true);
+
+    return sol_memmap_read_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_write_drange(const char *name, struct sol_drange *value)
+{
+    CREATE_BUFFER(value, false);
+
+    return sol_memmap_write_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_read_double(const char *name, double *value)
+{
+    CREATE_BUFFER(value, true);
+
+    return sol_memmap_read_raw(name, &buf);
+}
+
+static inline int
+sol_memmap_write_double(const char *name, double value)
+{
+    CREATE_BUFFER(&value, false);
+
+    return sol_memmap_write_raw(name, &buf);
```

Comment:
Since these implementations are the same for `memmap`, `fs`, and `efivars`, wouldn't it be possible to use something such as x-macros to implement these functions only once?